common/cah-model.cpp

```diff
@@ -0,0 +1,183 @@
+            "This process may take several minutes and requires special setup to get good results.\n"
+            "While it is working, the viewer will not be usable.\n\n";
+
+        if (fw_upgrade_needed)
+        {
+            std::string fw_upgrade_message = "Camera Accuracy Health requires a minimal FW version of " + min_fw_version +
+                "\n\nPlease update your firmware and try again. ";
+
+            message_text += fw_upgrade_message;
+        }
+        else
+        {
+            is_depth_streaming = std::any_of(dev_model.subdevices.begin(), dev_model.subdevices.end(), [](const std::shared_ptr<subdevice_model>& sm) { return sm->streaming && sm->s->as<depth_sensor>(); });
+            is_color_streaming = std::any_of(dev_model.subdevices.begin(), dev_model.subdevices.end(), [](const std::shared_ptr<subdevice_model>& sm) { return sm->streaming && sm->s->as<color_sensor>(); });
+
+            if (is_depth_streaming && is_color_streaming)
+            {
+                message_text += "Are you sure you want to continue?";
+            }
+            else
+            {
+                std::string stream_missing_message = "Camera Accuracy Health cannot be triggered : both depth & RGB streams must be active.";
+                message_text += stream_missing_message;
+            }
+
+        }
+
+        bool option_disabled = !is_depth_streaming || !is_color_streaming || fw_upgrade_needed;
+        if (yes_no_dialog("Camera Accuracy Health Trigger", message_text, yes_was_chosen, window, error_message, option_disabled))
+        {
+            if (yes_was_chosen)
+            {
+
+                auto itr = std::find_if(dev_model.subdevices.begin(), dev_model.subdevices.end(), [](std::shared_ptr<subdevice_model> sub)
+                {
+                    if (sub->s->as<depth_sensor>())
+                        return true;
+                    return false;
+                });
+
+
+                if (is_depth_streaming && is_color_streaming && itr != dev_model.subdevices.end())
+                {
+                    auto sd = *itr;
+                    sd->s->set_option(RS2_OPTION_TRIGGER_CAMERA_ACCURACY_HEALTH, 1.0f);
+                    device_calibration dev_cal(dev_model.dev);
+                    // Register AC status change callback
+                    if (!_registered_to_callback)
+                    {
+                        _registered_to_callback = true;
+                        _state = model_state_type::PROCESS_MODAL;
+                        // We switch to process state without a guarantee that the process really started,
+                        // Set a timeout to make sure if it is not started we will allow closing the window.
+                        _process_timeout.start(std::chrono::seconds(30)); 
+                        dev_cal.register_calibration_change_callback([&](rs2_calibration_status cal_status)
+                        {
+                            _calib_status = cal_status;
```

Comment:
Is it possible that the model object is no longer there when we get here?

Reply:
The device model is the one calling to that function with *this parameter.
I don't think it is possible that this situation can happen.

Reply:
But the callback can be called from AC even after the device model has been destroyed, no?

Reply:
Correct, will be fixed.
_calib_status will be a global static variable and will not be related to the cah_model class